src/Catalyst.Core.Lib/Dag/DagNode.cs

```diff
@@ -0,0 +1,379 @@
+            var ignore = links.ToLookup(link => link.Id);
+            var some = Links.Where(link => !ignore.Contains(link.Id));
+            return new DagNode(DataBytes, some, hashAlgorithm);
+        }
+
+        /// <summary>
+        ///     Writes the binary representation of the node to the specified <see cref="Stream" />.
+        /// </summary>
+        /// <param name="stream">
+        ///     The <see cref="Stream" /> to write to.
+        /// </param>
+        public void Write(Stream stream)
+        {
+            using (var cos = new CodedOutputStream(stream, true))
+            {
+                Write(cos);
+            }
+        }
+
+        /// <summary>
+        ///     Writes the binary representation of the node to the specified <see cref="CodedOutputStream" />.
+        /// </summary>
+        /// <param name="stream">
+        ///     The <see cref="CodedOutputStream" /> to write to.
+        /// </param>
+        public void Write(CodedOutputStream stream)
+        {
+            if (stream == null)
+                throw new ArgumentNullException("stream");
+
+            foreach (var link in Links.Select(l => new DagLink(l)))
+            {
+                using (var linkStream = new MemoryStream())
+                {
+                    link.Write(linkStream);
+                    var msg = linkStream.ToArray();
+                    stream.WriteTag(2, WireFormat.WireType.LengthDelimited);
+                    stream.WriteLength(msg.Length);
+                    stream.WriteSomeBytes(msg);
+                }
+            }
+
+            if (DataBytes.Length > 0)
+            {
+                stream.WriteTag(1, WireFormat.WireType.LengthDelimited);
+                stream.WriteLength(DataBytes.Length);
+                stream.WriteSomeBytes(DataBytes);
+            }
+        }
+
+        private void Read(Stream stream)
+        {
+            using (var cis = new CodedInputStream(stream, true))
+            {
+                Read(cis);
+            }
+        }
+
+        private void Read(CodedInputStream stream)
+        {
+            var links = new List<DagLink>();
+            var done = false;
+
+            while (!stream.IsAtEnd && !done)
+            {
+                var tag = stream.ReadTag();
+                switch (WireFormat.GetTagFieldNumber(tag))
+                {
+                    case 1:
+                        DataBytes = stream.ReadSomeBytes(stream.ReadLength());
+                        done = true;
+                        break;
+                    case 2:
+                        using (var ms = new MemoryStream(stream.ReadSomeBytes(stream.ReadLength())))
+                        {
+                            links.Add(new DagLink(ms));
+                        }
+
+                        break;
+                    default:
+                        throw new InvalidDataException("Unknown field number");
+                }
+            }
+
+            if (DataBytes == null)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)